ehcache-impl/src/main/java/org/ehcache/impl/internal/store/heap/SimpleBackend.java

```diff
@@ -122,26 +117,45 @@ public Collection<Map.Entry<K, OnHeapValueHolder<V>>> removeAllWithHash(int hash
 
   @Override
   public OnHeapValueHolder<V> remove(K key) {
-    return realMap.remove(key);
+    return mapHolder.getRealMap().remove(key);
   }
 
   @Override
   public OnHeapValueHolder<V> computeIfPresent(final K key, final BiFunction<K, OnHeapValueHolder<V>, OnHeapValueHolder<V>> computeFunction) {
-    return realMap.computeIfPresent(key, computeFunction);
+    return mapHolder.getRealMap().computeIfPresent(key, computeFunction);
   }
 
   @Override
   public OnHeapValueHolder<V> get(K key) {
-    return realMap.get(key);
+    return mapHolder.getRealMap().get(key);
   }
 
   @Override
   public OnHeapValueHolder<V> putIfAbsent(K key, OnHeapValueHolder<V> valueHolder) {
-    return realMap.putIfAbsent(key, valueHolder);
+    return mapHolder.getRealMap().putIfAbsent(key, valueHolder);
   }
 
   @Override
   public boolean replace(K key, OnHeapValueHolder<V> oldValue, OnHeapValueHolder<V> newValue) {
-    return realMap.replace(key, oldValue, newValue);
+    return mapHolder.getRealMap().replace(key, oldValue, newValue);
   }
+
+  static class MapWithByteSizeHolder<K, V> {
```

Comment:
This the right idea, but you should be able to make the `SimpleBackend` class the holder. That way the clear operation becomes a replacement of the SimpleBackend in the OnHeapStore layer (and the provider moves up a layer, and changes type appropriately).

Reply:
After a discussion got the context, and produce the fix by moving MapSupplier up a level at - `OnHeapStore.java`